phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala

```diff
@@ -0,0 +1,371 @@
+**     ________ ___   / /  ___      __ ____  PhantomJS support for Scala.js **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2017, LAMP/EPFL       **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    https://www.scala-js.org/      **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                                   **
+**                          |/____/                                         **
+\*                                                                          */
+
+package org.scalajs.jsenv.phantomjs
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Promise, Future}
+import scala.util.Try
+import scala.util.control.NonFatal
+
+import java.io._
+import java.net._
+import java.nio.file.{Files, StandardCopyOption}
+
+import org.scalajs.jsenv._
+
+import org.scalajs.io._
+import org.scalajs.io.URIUtils.fixFileURI
+import org.scalajs.io.JSUtils.escapeJS
+
+private final class ComRun(jettyClassLoader: ClassLoader,
+    runConfig: RunConfig, onMessage: String => Unit,
+    startRun: VirtualBinaryFile => JSRun)
+    extends JSComRun {
+
+  import ComRun._
+
+  import runConfig.logger
+
+  // Maybe this should be configurable
+  private implicit val ec = ExecutionContext.global
+
+  private var state: State = WaitingForServerToStart(Nil)
+  private val connectionClosed = Promise[Unit]()
+
+  private[this] val fragmentsBuf = new StringBuilder
+
+  private val promise = Promise[Unit]()
+
+  def future: Future[Unit] = promise.future
+
+  private def loadMgr(): WebsocketManager = {
+    val loader =
+      if (jettyClassLoader != null) jettyClassLoader
+      else getClass().getClassLoader()
+
+    val clazz = loader.loadClass(
+        "org.scalajs.jsenv.phantomjs.JettyWebsocketManager")
+
+    val ctors = clazz.getConstructors()
+    assert(ctors.length == 1, "JettyWebsocketManager may only have one ctor")
+
+    val listener = new WebsocketListener {
+      def onRunning(): Unit = onServerRunning()
+      def onOpen(): Unit = onConnectionOpened()
+      def onClose(): Unit = onConnectionClosed()
+      def onMessage(msg: String): Unit = receiveFrag(msg)
+      def log(msg: String): Unit = logger.debug(s"PhantomJS WS Jetty: $msg")
+    }
+
+    val mgr = ctors.head.newInstance(listener)
+
+    mgr.asInstanceOf[WebsocketManager]
+  }
+
+  private val mgr: WebsocketManager = loadMgr()
+
+  // Constructor
+  mgr.start()
+
+  private def onServerRunning(): Unit = synchronized {
+    state match {
+      case WaitingForServerToStart(sendQueue) =>
+        state = AwaitingConnection(sendQueue)
+        val comSetup = makeComSetupFile(mgr.localPort)
+        val underlyingRun = startRun(comSetup)
+
+        underlyingRun.future.onComplete { result =>
+          underlyingRun.close()
+          onUnderlyingRunTerminated(result)
+        }
+
+      case Closing =>
+        // Ignore
+
+      case AwaitingConnection(_) | Connected | NotConnectedAnymore =>
```

Comment:
`NotConnectedAnymore` seems to never be set. Am I missing something?

Reply:
Indeed. Not sure how that happened. I'll have to revisit when I get the chance. (I'm on vacation until August 13.)

Reply:
Taken care of. See the new commit (to be squashed later).